service/storage_proxy.cc

```diff
@@ -159,6 +165,85 @@ class storage_proxy::remote {
     netw::connection_drop_slot_t _connection_dropped;
     netw::connection_drop_registration_t _condrop_registration;
 
+    class request_info {
+    public:
+        virtual future<schema_ptr> get_schema() const = 0;
+        virtual dht::token get_token(const schema& schema) const = 0;
+    };
+
+    struct read_request_info: public request_info {
+        schema_ptr _schema;
+        const dht::partition_range& _range;
+    public:
+        read_request_info(schema_ptr schema, const dht::partition_range& range)
+            : _schema(schema)
+            , _range(range)
+        {
+        }
+        virtual future<schema_ptr> get_schema() const override {
+            return make_ready_future<schema_ptr>(_schema);
+        }
+        virtual dht::token get_token(const schema&) const override {
+            return end_token(_range);
+        }
+    };
+
+    class write_request_info: public request_info {
+        remote& _remote;
+        table_schema_version _schema_version;
+        netw::msg_addr _msg_from;
+        partition_key_view _partition_key;
+    public:
+        write_request_info(remote& remote, table_schema_version schema_version, netw::msg_addr msg_from,
+            partition_key_view _partition_key)
+            : _remote(remote)
+            , _schema_version(schema_version)
+            , _msg_from(msg_from)
+            , _partition_key(std::move(_partition_key))
+        {
+        }
+        virtual future<schema_ptr> get_schema() const override {
+            return _remote.get_schema_for_write(_schema_version, _msg_from);
+        }
+        virtual dht::token get_token(const schema& schema) const override {
+            return schema.get_partitioner().get_token(schema, _partition_key);
+        }
+    };
+
+    future<std::optional<replica::stale_topology_exception>> try_apply_fence(fencing_token fence,
+        lowres_clock::time_point timeout,
+        const request_info& request_info)
+    {
+        while (true) {
+            const auto current_version = _sp._shared_token_metadata.get()->get_topology_version();
```

Comment:
We need to capture the topology when the request starts (before this function) and use it everywhere. Otherwise different paths can use different topologies.

See 87b7e7ff9ca1ff6524c5f67a43da56693004d2cb for similar idea (but coordinator side)

Reply:
This function is used on non-coordinator slave replicas, where we should either do local write or forward the write request to other dc-local replicas. IIUC, the topology is used only on coordinator to detect which nodes are responsible for processing the request, and in those places the topology is already captured.

In this specific loop in `fence.topology_version < current_version` case `token_metadata` can actually change after `co_await get_schema`. And this is probably a bug, since it can change in such a way that the `token_owned` variable would has a different value. I will fix this.

Reply:
I think we should capture the topology on replicas too. This allows two things:

1. the replica checks that its topology is compatible with the sender topology
2. the replica holds the topology for the duration of the request, allowing the replica to answer with authority to the request "switch the topology to this here version and tell me when there are no more users of your older version" (by waiting for the reference count of the older version to drop to zero)

Reply:
@avikivity I do not understand what is your idea of how fencing suppose to work. You clearly have something in mind that probably does not match what others have in mind. What I do not understand is why would the code, which purpose is to waits for a certain version of topology, capture the topology. If it would it will always check the same version of the topology and loop forever.

Reply:
First, I don't think we should wait for topology. The coordinator should send a topology it knows the replica has. Fencing should eliminate outdated requests with very old topology from being accepted. It's not a tool for synchronizing topology, that's a job for the control plane.

We need to capture the topology on the replica to prevent the following race.
1. write starts with correct topology
2. while the write waits, the topology changes
3. write ends up using incorrect topology

Reply:
> > First, I don't think we should wait for topology. The coordinator should send a topology it knows the replica has.
> 
> What does it mean "The coordinator should send a topology it knows the replica has."? First the coordinator does not know what topology replica has

Why not? If we have a topology change coordinator, then it can update every node with the state of every reachable node. The topology change coordinator is guaranteed to reach a majority of the nodes. 

We won't know the state of unreachable nodes, but they're unreachable anyway.

> and second how can it "send a topology"?

The version of the topology.

> 
> > Fencing should eliminate outdated requests with very old topology from being accepted.
> 
> That what the patch is doing exactly. But what if this node with old topology get newer request. What in your opinion should it do?

It should also allow the node to report when it stopped using a topology version. This is why I want the topology ptr to be captured and held.

What I don't want it to do is wait for the topology to catch up. If we wait (or initiate topology synchronization), everything works, but we introduced large waits on the control plane. Instead, the control plane should only allow using the new topology when it is sure everyone is ready for it.

> And in your opinion it cannot happen explain how.



> 
> > It's not a tool for synchronizing topology, that's a job for the control plane.
> 
> There is no topology synchronization here. There is a wait for control plane to do the sync.
> 
> > We need to capture the topology on the replica to prevent the following race.
> > ```
> > 1. write starts with correct topology
> > 
> > 2. while the write waits, the topology changes
> > 
> > 3. write ends up using incorrect topology
> > ```
> 
> How capturing the topology on the replica helps it any? During write on a replica we do not consult the topology (well except the code we add here to check its version). Replica wrights/reads assume that the coordinator uses correct topology and fence requests that belongs to the wrong one as the code here does.


We implicitly use the topology by accepting the read or write. For a read, if we send a result after the node is no longer part of the replica set, we send an incorrect result. For a write, if we acknowledge after the node is no longer part of a replica set, we lead the coordinator to an incorrect conclusion about whether the consistency level was reached or not.

Reply:
> > > First, I don't think we should wait for topology. The coordinator should send a topology it knows the replica has.
> > 
> > 
> > What does it mean "The coordinator should send a topology it knows the replica has."? First the coordinator does not know what topology replica has
> 
> Why not? If we have a topology change coordinator, then it can update every node with the state of every reachable node. The topology change coordinator is guaranteed to reach a majority of the nodes.

I do not know where to start. First RAFT does not give you this "now all nodes are exactly synced" property that you seams to want. Second a request coordinator is not the same as topology change coordinator. Last but not least a quorum of all the nodes in group0 are not the same as a quorum for a particular request. In fact a request may contact only nodes which were not in the quorum for the latest topology change at all. 

> 
> We won't know the state of unreachable nodes, but they're unreachable anyway.

They may be unreachable for a topology change coordinator, but it does not mean it cannot be alive and be a coordinator for cql requests.
  
> 
> > and second how can it "send a topology"?
> 
> The version of the topology.

That what the patch does, no?

> 
> > > Fencing should eliminate outdated requests with very old topology from being accepted.
> > 
> > 
> > That what the patch is doing exactly. But what if this node with old topology get newer request. What in your opinion should it do?
> 
> It should also allow the node to report when it stopped using a topology version. This is why I want the topology ptr to be captured and held.

As I wrote below replicas do not use topology.

> 
> What I don't want it to do is wait for the topology to catch up. If we wait (or initiate topology synchronization), everything works, but we introduced large waits on the control plane. Instead, the control plane should only allow using the new topology when it is sure everyone is ready for it.

That what control plane will do as well. Current topology code have read/write barrier where it contacts each node and asks it to wait for all current requests to complete (the code that waits is missing right now but eventually it will be there). Only after that we will move to the next version.  This will take to the minimum the number of requests with older version. But we cannot stop all new requests to come in while we synchronize the cluster and while cluster is been synchronized there can be a request with newer version going to a node with older version. It will
happen only while control plane doing the sync. Why is it wrong to wait for it here? Do you prefer to block all requests on a coordinator while it happens? 

> 
> > And in your opinion it cannot happen explain how.
> 
> > > It's not a tool for synchronizing topology, that's a job for the control plane.
> > 
> > 
> > There is no topology synchronization here. There is a wait for control plane to do the sync.
> > > We need to capture the topology on the replica to prevent the following race.
> > > ```
> > > 1. write starts with correct topology
> > > 
> > > 2. while the write waits, the topology changes
> > > 
> > > 3. write ends up using incorrect topology
> > > ```
> > 
> > 
> > How capturing the topology on the replica helps it any? During write on a replica we do not consult the topology (well except the code we add here to check its version). Replica wrights/reads assume that the coordinator uses correct topology and fence requests that belongs to the wrong one as the code here does.
> 
> We implicitly use the topology by accepting the read or write. For a read, if we send a result after the node is no longer part of the replica set, we send an incorrect result. For a write, if we acknowledge after the node is no longer part of a replica set, we lead the coordinator to an incorrect conclusion about whether the consistency level was reached or not.

Yes, and that is why we have all this dance with pending/ring_delay in gossiper and those state the range comes through while streaming with raft. But in the end the one who makes sure that correct topology is used today is a request coordinator only, so you just bagged my question about what exactly a replica need to capture topology for.

This patch is actually introduces the first usage of the topology on read/write path for fencing, which basically make sure that the coordinator uses the same version of the topology as the replica, and, if they are different, it checks that the data placement is compatible.  So where exactly it should hold the topology here and for how long?

Reply:
@avikivity @gleb-cloudius it seems [we discussed](https://github.com/gleb-cloudius/scylla/pull/1#discussion_r1048310538) and agreed on something similar to what @avikivity is proposing here and below:

1. `raft_apply(tokens_state=ReadWrite, version=current_version+1)` accept current_version+1 and current_version, store this fact (about allowing previous version) only in memory. `ReadWrite` here means data is written to both old and new replicas and is read from new replicas.
2. `wait_all_completed(timeout=several seconds)`, run on all nodes, read_barrier on each (meaning raft state from 1 is applied), wait for outgoing RPCs with current_version to finish
3. `fence`, run on all nodes, start to fence old RPCs. Turn the in-memory flag and start throwing errors on current_version.
4. `raft_apply(tokens_state=owned)`. Here we need that either there are no nodes using old topology for sending RPCs _or_ no nodes accepting old topology (these requirements are pretty much the same in terms of the end result). If both 2 and 3 above have run successfully on all nodes, we have both. 

step 2 gives current requests the change to finish gracefully, but it has a timeout so no guarantee that all the requests with current_version are finished. That's why we need step 3 where we explicitly turn on fencing for all nodes. And this guarantees that during step 4 no requests will be fulfilled based on the old topology.

Also we discussed that steps 2 and 3 can be run not on all nodes, but only on a subset of 'affected' nodes - those stored tokens before or those that will store them after the topology change. But this is an enhancement which can be done later.

This patch doesn't implement all the means needed for 1-3. We can discuss/settle on this, and I will add them, in this PR or in the follow-ups.

Reply:
What about the wait for topology version? That's the part I don't like. I don't like it because it injects control-plane latency (propagating the topology) into the data plane.

The sender should be certain that downstream has a new-enough topology. For sure it cannot be achieved if we have network partitioning, but when network partitioning happens our latency is equal to the partition heal latency.

For writes the situation is more complicated, for example for LOCAL_QUORUM writes we need to ensure only that two local nodes have the updated topology, and don't need to care about remove datacenters, even though we do want to write to all replicas.

Reply:
> What about the wait for topology version? That's the part I don't like. I don't like it because it injects control-plane latency (propagating the topology) into the data plane.

This is irrational fear because it already does :) We already sync schema on read/write which is a "control-plane" operation done in the data plane.
 
> 
> The sender should be certain that downstream has a new-enough topology. For sure it cannot be achieved if we have network partitioning, but when network partitioning happens our latency is equal to the partition heal latency.
> 

I do not how it can be achieved without stopping the traffic.

Reply:
> > What about the wait for topology version? That's the part I don't like. I don't like it because it injects control-plane latency (propagating the topology) into the data plane.
> 
> This is irrational fear because it already does :) We already sync schema on read/write which is a "control-plane" operation done in the data plane.

I don't like it and I think we should treat the schema the same way.


> 
> > The sender should be certain that downstream has a new-enough topology. For sure it cannot be achieved if we have network partitioning, but when network partitioning happens our latency is equal to the partition heal latency.
> 
> I do not how it can be achieved without stopping the traffic.

I don't understand why the traffic has to be stopped. The topology state coordinator waits until all reachable nodes have the new topology, then instructs them to start using the new topology. If the topology state coordinator crashes, the new one does the same. Meanwhile the traffic flows, first with the old version, then with a mix of the old and new versions (as nodes are instructed to switch), and then just the new version.

Reply:
@avikivity Gleb was able to explain your concerns in the end, but I must note that they are complementary to what is going on here.
When moving the cluster topology forward we have the following *tools*:
* reject new requests *at request coordinator* if the node which received the request has stale topology
* reject new mutations/reads *at replicas* if the coordinator has the stale topology
* retry mutations/reads *at request coordinator* after learning its topology is outdated and refreshing a topology
* drop requests *at request coordinator* if it learns that its topology is outdated
* gradually roll out a topology change over to cluster membres using green/blue deployment principles at *topology coordinator* - the topology coordinator needs a way to wait for writes to stop
* cliff roll out a topology change over to cluster members knowing fencing off some (presumably dead) members of the cluster

Please note that the choice of actions of topology coordinator is orthogonal to the  choice of actions of replicas/request coordinators. E.g. the topology coordinator can choose to instruct replicas to fence off old writes, or softly wait for old writes, and replicas may choose to drop client requests or retry them - depending on the specific scenario. 

This patch is about fencing RPC, and there is nothing wrong in it implementing retries - if coordinator can wait for old writes to naturally die out, the retires won't be needed. 

Please note that it can't in a general case, because of asymmetric network partitioning scenarios replicas can have a never ending stream of writes using the old topologies. So at some point, under some conditions, the topology coordinator will have to rely on a timeout. E.g. imagine you aborting a failed node addition, and during abort one of the node is disconnected from the cluster. You can't wait forever, you have to fence off the node. Once you fenced it off, you may choose what to do on that node when it gets connected: do you want to drop all requests from that node until it refreshes the topology, or do you want it to be able to retry? Both options have to be on the table.

Reply:
The main point I am trying to make: how the coordinator decides to roll out the topology is outside the scope of this patch. You believe that it should be rolled out in a smooth manner, that's fine, and hence retires will never be necessary  - that's less obvious. It's clear they can be harmful when they occur, since they lead to overload/accumulating effects. But we already perform retries of standard reads and writes, not to mention LWT, so have to deal with overload handling anyway.

Reply:
@avikivity I dropped waiting for the topology from `try_apply_fence`, throw instead. Do we need anything else from this tread?

Reply:
I lost the context, I'll have to restart my review.

What's important to me:
 - no mingling of the control and data planes
 - no errors in normal operation